crates/nri-resctrl-plugin/src/lib.rs

```diff
@@ -332,6 +341,196 @@ impl<P: FsProvider> ResctrlPlugin<P> {
+            }
+            Err(resctrl::Error::Capacity { .. }) => Err(resctrl::Error::Capacity {
+                source: std::io::Error::from_raw_os_error(libc::ENOSPC),
+            }),
+            Err(e) => Err(e),
+        }
+    }
+
+    /// Retry reconciling a single container if its pod group exists.
+    /// Emits AddOrUpdate only if reconciled count is incremented.
+    pub(crate) fn retry_container_reconcile(
+        &self,
+        container_id: &str,
+    ) -> resctrl::Result<ContainerSyncState> {
+        use std::io;
+        use std::path::PathBuf;
+
+        // Snapshot under lock: group path, cgroup path, passes, current state
+        let (group_path, cgroup_path, pod_uid, _current_state, passes) = {
+            let st = self.state.lock().unwrap();
+            let cst = st
+                .containers
+                .get(container_id)
+                .ok_or_else(|| resctrl::Error::Io {
+                    path: PathBuf::from("<container>"),
+                    source: io::Error::from_raw_os_error(libc::ENOENT),
+                })?;
+            if cst.state == ContainerSyncState::NoPod {
+                return Ok(ContainerSyncState::NoPod);
+            }
+            let ps = st
+                .pods
+                .get(&cst.pod_uid)
+                .ok_or_else(|| resctrl::Error::Io {
+                    path: PathBuf::from("<pod>"),
+                    source: io::Error::from_raw_os_error(libc::ENOENT),
+                })?;
+            let gp = match &ps.group_state {
+                ResctrlGroupState::Exists(p) => p.clone(),
+                _ => return Ok(cst.state),
+            };
+            (
+                gp,
+                cst.cgroup_path.clone(),
+                cst.pod_uid.clone(),
+                cst.state,
+                self.cfg.max_reconcile_passes,
+            )
+        };
+
+        // Perform reconcile outside the lock
+        let pid_source = self.pid_source.clone();
+        let pid_resolver =
+            move || -> resctrl::Result<Vec<i32>> { pid_source.pids_for_path(&cgroup_path) };
+        let res = self
+            .resctrl
+            .reconcile_group(&group_path, pid_resolver, passes)?;
+        let new_state = if res.missing == 0 {
+            ContainerSyncState::Reconciled
+        } else {
+            ContainerSyncState::Partial
+        };
+
+        // Re-acquire lock and update counters/state conditionally
+        let mut st = self.state.lock().unwrap();
+        // Track whether we improved state
+        let mut did_reconcile = false;
+        let ret_state = if let Some(cst) = st.containers.get_mut(container_id) {
+            if cst.state == ContainerSyncState::Partial
+                && new_state == ContainerSyncState::Reconciled
+            {
+                cst.state = ContainerSyncState::Reconciled;
+                did_reconcile = true;
+            }
+            cst.state
+        } else {
+            return Err(resctrl::Error::Io {
```

Comment:
if we add an error for container not found (comment above), this should also be that error...